sanic/headers.py

```diff
@@ -0,0 +1,19 @@
+import re
+
+
+token, quoted = r"([\w!#$%&'*+\-.^_`|~]+)", r'"((?:[^"]|\\")*)"'
+parameter = re.compile(fr";\s*{token}=(?:{token}|{quoted})", re.ASCII)
+
+
+def parse_options_header(value: str):
+    """Parse HTTP header values of Content-Type format."""
+    pos = value.find(";")
+    if pos == -1:
+        options = {}
+    else:
+        options = {
+            m.group(1).lower(): m.group(2) or m.group(3).replace('\\"', '"')
+            for m in parameter.finditer(value[pos:])
+        }
+        value = value[:pos]
+    return value.strip().lower(), options
```

Comment:
we should add the existing cpython tests to the test suite for sanic, here they are converted to pytest:

```import pytest
from sanic.headers import parse_options_header


@pytest.mark.parametrize(
    "input, expected",
    [
        ("text/plain", ("text/plain", {})),
        ("text/vnd.just.made.this.up ; ", ("text/vnd.just.made.this.up", {})),
        ("text/plain;charset=us-ascii", ("text/plain", {"charset": "us-ascii"})),
        ('text/plain ; charset="us-ascii"', ("text/plain", {"charset": "us-ascii"})),
        (
            'text/plain ; charset="us-ascii"; another=opt',
            ("text/plain", {"charset": "us-ascii", "another": "opt"})
        ),
        (
            'attachment; filename="silly.txt"',
            ("attachment", {"filename": "silly.txt"})
        ),
        (
            'attachment; filename="strange;name"',
            ("attachment", {"filename": "strange;name"})
        ),
        (
            'attachment; filename="strange;name";size=123;',
            ("attachment", {"filename": "strange;name", "size": "123"})
        ),
        (
            'form-data; name="files"; filename="fo\\"o;bar"',
            ("form-data", {"name": "files", "filename": 'fo"o;bar'})
        ),
    ]
)
def test_parse_headers(input, expected):
    assert parse_options_header(input) == expected

```

the last test misses. I think the issue is the `quoted` string should be a normal string instead of a raw

Reply:
A long time ago browsers didn't escape and that became the common practice; to deal with all the existing broken CGI scripts, modern browsers opt to mark it as %22 instead of attempting to use `\"`. Sanic's form decoding code should probably decode these as well (currently it doesn't and the %-escape is in the returned filename too). I believe it is in order to modify or remove that last test.

Reply:
gotcha, I'm fine with removing that last test then, but we should probably remove the `\\"` part of the `quoted` token then so its are clear we explicitly aren't observing quoting

Reply:
It is included in the RFCs and probably actually used in some other headers, so I'd prefer to keep it if it doesn't cause trouble.

Reply:
Okay, I just tried with a file named `test\` (on unix)
``
`
Content-Disposition: form-data; name="foo"; filename="test\"
```
Looks like we need to remove that escape.